internal/controller/scheduler/batch_buildkite_job_checker.go

```diff
@@ -0,0 +1,180 @@
+package scheduler
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/buildkite/agent-stack-k8s/v2/api"
+	"github.com/google/uuid"
+	"go.uber.org/zap"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+// BuildkiteJobChecker monitors Buildkite jobs for cancellation state changes.
+// It operate on individual jobs basis, relying on legacy Agent API.
+// In the future version, we may get rid of it.
+type BatchBuildkiteJobChecker struct {
+	logger      *zap.Logger
+	agentClient *api.AgentClient
+	k8s         kubernetes.Interface
+
+	// The job cancel checkers query the job state every so often.
+	jobCancelCheckerInterval time.Duration
+
+	// Store jobs that we will check against.
+	checkingJobsMu sync.Mutex
+	checkingJobs   map[uuid.UUID]metav1.ObjectMeta
+}
+
+// NewBatchBuildkiteJobChecker creates a new Buildkite job checker.
+// Unlike the old legacy BuildkiteJobChecker, this checker check all pending jobs together,
+// relying on the new Stack API.
+func NewBatchBuildkiteJobChecker(
+	logger *zap.Logger,
+	agentClient *api.AgentClient,
+	k8s kubernetes.Interface,
+	interval time.Duration,
+) *BatchBuildkiteJobChecker {
+	return &BatchBuildkiteJobChecker{
+		logger:                   logger,
+		agentClient:              agentClient,
+		k8s:                      k8s,
+		jobCancelCheckerInterval: interval,
+		checkingJobs:             make(map[uuid.UUID]metav1.ObjectMeta),
+	}
+}
+
+// Designed to run in a goroutine. It starts a loop to periodically check job states for all jobs that it manages.
+func (c *BatchBuildkiteJobChecker) StartChecking(ctx context.Context) {
+	c.logger.Debug("Starting batch job state checker")
+	defer c.logger.Debug("Stopped batch job state checker")
```

Comment:
i reckon if this is always intended to be run in a goroutine, it should just start that goroutine itself rather than relying on the caller